polyfills/promise.js

```diff
@@ -9,6 +9,9 @@
  *    - Doesn't handle errors from core operations, e.g. out-of-memory or
  *      internal error when queueing/running jobs.  These are implementation
  *      defined for the most part.
+ *    - Unhandled rejection 'reject' and 'handle' events don't have the same
+ *      ordering as in Node.js at present.  Unhandled rejection behavior also
+ *      hasn't been checked against latest ES specifications.
```

Comment:
I don’t understand how this can be; by definition you only get a `handle` event for a promise which previously generated a `reject` event, so it’s impossible to have them in the opposite  order I would think?

Reply:
The comment refers to more subtle ordering guarantees, like what's the ordering between Promise job queue entries vs. unhandled rejection calls. For example, for https://github.com/svaarala/duktape/blob/68f1b824fa287f075dda4d0a34ae1ebb1bd6586d/tests/ecmascript/test-bi-promise-unhandled-difftick-1.js Duktape prints:
```
tick 1
unhandled/reject: P
unhandled/reject: Q
unhandled/reject: R
tick 2
nop called
nop called
unhandled/handle: P
unhandled/handle: Q
unhandled/handle: R
unhandled/reject: R2
tick 3
tick 4
```

and Node.js prints:
```
tick 1
unhandled/reject: P
unhandled/reject: Q
unhandled/reject: R
tick 2
unhandled/handle: P
unhandled/handle: Q
unhandled/handle: R
nop called
nop called
unhandled/reject: R2
tick 3
tick 4
```

This is unlikely to matter in practice, and I'll refine the polyfill later on w.r.t. ordering.

Reply:
Some other minor ordering issues:
- Relationship between events for multiple Promises dealt with in one tick. If P and Q are unhandled, are the events guaranteed to happen in a specific order? For instance, is it acceptable to emit `reject` for P,Q but `handle` for Q,P? Obviously consistent ordering is preferable, but is it required?
- Relationship between unhandled rejection callbacks and Promise jobs. For example, an unhandled rejection callback may cause a Promise job, or more unhandled rejections to be queued. What ordering guarantees are made for them? For Promise jobs the order is fully deterministic as far as I can tell, but I'm not sure about unhandled rejections because ES doesn't specify any ordering guarantees for different queue types.
- If a Promise first gets a `reject` and then gets handled, is there a guarantee that the `handle` event happens before the handler itself is called (or vice versa)? Node.js seems to do so, but I'm not sure this is required.

I'm not trying to address these at present, just to point out there are several ordering possibilities and levels of guarantees one can try to provide.

Reply:
The `handle` event is associated with the `.then()` call (or alias, e.g. `.catch()`) so I would always expect it to go into the queue before the handler call itself.  The PromiseRejectionTracker callback is not allowed to throw so I think the expectation is that the engine calls it synchronously.

Reply:
Is there a requirement for `reject` and `handle` to coexist in the Promise job queue?

I first thought Node.js queues `reject` and `handle` events in the Promise job queue, but at least in some tests (very quick I admit) that doesn't seem to be the case.

Reply:
Re: Node.js behavior, suppose that two Promises, P and Q, have been emitted a `reject` event.

You then handle the promises e.g. as:
```
P.catch(nop);
Q.catch(nop);
```

If everything went through the job queue, one would expect to see:
- `handle` for P
- nop called for P
- `handle` for Q
- nop called for Q

But, this doesn't seem to happen with Node.js. Instead:
- `handle` for P
- `handle` for Q
- nop called for P
- nop called for Q

So it seems that pending `handle` events are handled before the "tick", while `reject` events are handled after the "tick".

But I haven't had much time to test Node.js or browser behavior further, so I won't make changes in the polyfill until there's a better idea of what is (a) actually required, (b) desirable in practice.

Reply:
Here's that Node.js test which to me implies the `reject` and `handle` events don't coexist in the Promise job queue:

```
process.on('unhandledRejection', function (reason, promise) {
    console.log('reject:', promise.name);
});
process.on('rejectionHandled', function (promise) {
    console.log('handle:', promise.name);
});
var P = Promise.reject(123); P.name = 'P';
var Q = Promise.reject(234); Q.name = 'Q';

setTimeout(function () {
    P.catch(function () { console.log('P.catch'); });
    Q.catch(function () { console.log('Q.catch'); });
}, 0);
```

Output is:
```
reject: P
reject: Q
handle: P
handle: Q
P.catch
Q.catch
```

Reply:
As a side note, it also implies that expired timers (a zero timer is always expired if one processes all event loops to completion) *do not* count towards the "tick" definition for unhandled rejections. Otherwise no `reject` or `handle` events would be emitted.